File: dotnet/Functions/SendOtp.cs
using System.Text.Json;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Extensions.Logging;

namespace Cyot.Otp;

// HTTP trigger: POST /api/SendOtp — the SAS → CYOT delivery endpoint. Validates the Entra token, parses
// the cleartext routing envelope, decrypts the JWE delivery context (PII lives there), dispatches to the
// provider, and echoes the nonce to prove decryption. Privacy: phone and OTP code are never logged or
// returned; the response body is the minimal CyotEndpointResponse.
public sealed class SendOtp
{
    private readonly DispatchEngine _engine;
    private readonly TokenValidator _tokens;
    private readonly JweDecryptor _decryptor;
    private readonly ILogger<SendOtp> _log;

    public SendOtp(DispatchEngine engine, TokenValidator tokens, JweDecryptor decryptor, ILogger<SendOtp> log)
    {
        _engine = engine;
        _tokens = tokens;
        _decryptor = decryptor;
        _log = log;
    }

    [Function("SendOtp")]
    public async Task<IActionResult> Run(
        [HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "SendOtp")] HttpRequest req)
    {
        var requestId = Guid.NewGuid().ToString("n");
        var clientRequestId = req.Headers["x-ms-client-request-id"].FirstOrDefault() ?? requestId;
        var headerCorrelationId = req.Headers["x-ms-correlation-id"].FirstOrDefault();

        var auth = await _tokens.ValidateAsync(req.Headers.Authorization.FirstOrDefault());
        if (!auth.Ok)
        {
            _log.LogWarning("[AUTH_ERROR] requestId={RequestId} reason={Reason}", requestId, auth.Reason);
            return new ObjectResult(new { error = "unauthorized", reason = auth.Reason, requestId }) { StatusCode = 401 };
        }

        JsonElement payload;
        try
        {
            using var doc = await JsonDocument.ParseAsync(req.Body);
            payload = doc.RootElement.Clone();
        }
        catch
        {
            _log.LogWarning("[ERROR] requestId={RequestId} invalid JSON body", requestId);
            return new BadRequestObjectResult(new { error = "bad_request", reason = "invalid JSON body", requestId });
        }

        var (envelope, envelopeError) = EnvelopeParser.Parse(payload);
        if (envelopeError is not null)
        {
            _log.LogWarning("[VALIDATION_ERROR] requestId={RequestId} {Reason}", requestId, envelopeError);
            return new BadRequestObjectResult(new { error = "bad_request", reason = envelopeError, requestId });
        }

        var correlationId = envelope!.CorrelationId ?? headerCorrelationId ?? requestId;

        CyotDeliveryContext context;
        try
        {
            context = _decryptor.Decrypt(envelope.EncryptedDeliveryContext);
        }
        catch (Exception ex)
        {
            _log.LogWarning("[DECRYPT_ERROR] requestId={RequestId} correlationId={CorrelationId} reason={Reason}", requestId, correlationId, ex.Message);
            return new ObjectResult(new { error = "decryption_failed", correlationId, requestId }) { StatusCode = 400 };
        }

        if (string.IsNullOrEmpty(context.Nonce) || string.IsNullOrEmpty(context.PhoneNumber) || string.IsNullOrEmpty(context.Message))
        {
            _log.LogWarning("[VALIDATION_ERROR] requestId={RequestId} correlationId={CorrelationId} incomplete delivery context", requestId, correlationId);
            return new ObjectResult(new { error = "bad_request", reason = "incomplete delivery context", correlationId, requestId }) { StatusCode = 400 };
        }

        var evaluation = envelope.Mode == EnvelopeParser.ModeEvaluation;
        var channel = EnvelopeParser.ChannelName(envelope.Channel)!;

        // Respect ttlSeconds: don't start a live delivery for an already-expired passcode (contract §7).
        if (!evaluation && envelope.TtlSeconds is <= 0)
        {
            _log.LogWarning("[EXPIRED] requestId={RequestId} correlationId={CorrelationId} ttl={Ttl}", requestId, correlationId, envelope.TtlSeconds);
            return new ObjectResult(new { error = "request_expired", correlationId, requestId }) { StatusCode = 400 };
        }

        _log.LogInformation(
            "[SENDOTP] requestId={RequestId} caller={Caller} type={Type} tenant={Tenant} correlationId={CorrelationId} channel={Channel} mode={Mode} ttl={Ttl} phone=present message=present risk={Risk}",
            requestId, auth.CallerObjectId ?? "n/a", envelope.Type ?? "n/a", envelope.TenantId ?? "n/a", correlationId,
            envelope.Channel, envelope.Mode, envelope.TtlSeconds?.ToString() ?? "n/a", context.RiskContext.HasValue ? "present" : "absent");

        var dispatch = new DispatchRequest(
            Destination: context.PhoneNumber!,
            Message: context.Message,
            Channel: channel,
            MessageId: clientRequestId,
            CorrelationId: correlationId,
            Locale: context.Locale);

        try
        {
            var result = await _engine.DispatchAsync(dispatch, null, evaluation, requestId, _log);
            // Contract: acceptance is 202 Accepted (async delivery); the engine signals acceptance as 200.
            var accepted = result.HttpStatus == 200;
            return new ObjectResult(new { nonce = context.Nonce, correlationId, providerStatus = accepted ? "accepted" : "failed" })
                { StatusCode = accepted ? 202 : result.HttpStatus };
        }
        catch (Exception ex)
        {
            _log.LogError("[EXCEPTION] requestId={RequestId} error={Error}", requestId, ex.Message);
            return new ObjectResult(new { nonce = context.Nonce, correlationId, providerStatus = "failed" }) { StatusCode = 500 };
        }
    }
}

File: javascript/src/functions/SendOtp.js
// <copyright file="SendOtp.js" company="Microsoft Corporation">
// Copyright (c) Microsoft Corporation. All rights reserved.
// </copyright>

'use strict';

// POST /api/SendOtp — the SAS → External Phone Provider delivery endpoint. Validates the caller, parses
// the cleartext routing envelope, decrypts the JWE delivery context (PII lives there), dispatches to the
// provider, and echoes the nonce to prove decryption.
//
// Every line starts with [EPP], so a whole delivery can be pulled out of a noisy log with one filter:
//   Log stream           : look for [EPP]
//   Application Insights : traces | where message startswith "[EPP]" | order by timestamp asc

const { app } = require('@azure/functions');
const crypto = require('crypto');
const { validateToken } = require('./security');
const { dispatchOtp } = require('./dispatch');
const { parseEnvelope, decryptDeliveryContext, contextToDispatch, MODE } = require('./cyot');
const { readConfig, missingSettings } = require('./config');

const TAG = '[EPP]';

// Easy Auth has already validated the token by the time this runs; this only records which identity
// actually arrived, which is the first thing worth knowing when a delivery is refused.
function readCallerAppId(request) {
    const encoded = request.headers.get('x-ms-client-principal');
    if (!encoded) return undefined;
    try {
        const principal = JSON.parse(Buffer.from(encoded, 'base64').toString('utf8'));
        const claim = (principal.claims || []).find((c) => c.typ === 'appid' || c.typ === 'azp');
        return claim && claim.val;
    } catch {
        return undefined;
    }
}

// Lifts the passcode out of the rendered sentence, purely so it is easy to eyeball in the log.
function extractPasscode(message) {
    const match = /\b\d{4,8}\b/.exec(message || '');
    return match ? match[0] : undefined;
}

const pad = (label) => label.padEnd(18, ' ');

// The handler deliberately does not await the provider, so tests need a handle on the send it started.
let pendingDelivery = Promise.resolve();
const whenDelivered = () => pendingDelivery;

// Microsoft allows 3.2 s for the whole call, so the provider is called after the response.
function deliverInBackground(dispatch, envelope, evaluation, config, context, requestId) {
    pendingDelivery = dispatchOtp(dispatch, {
        tenantId: envelope.tenantId,
        requestProvider: config.provider.name || undefined,
        shutter: evaluation,
        context,
        requestId,
    })
        .then(({ httpStatus }) => {
            context.log(`${TAG} provider result   : httpStatus=${httpStatus}`);
        })
        .catch((deliveryError) => {
            (context.error || context.log).call(context, `${TAG} provider delivery failed: ${deliveryError.message}`);
        });
    return pendingDelivery;
}

app.http('SendOtp', {
    methods: ['POST'],
    authLevel: 'anonymous', // Easy Auth is the gate; EPP_REQUIRE_AUTH adds in-process token validation.
    handler: async (request, context) => {
        const started = Date.now();
        const config = readConfig();
        const log = (label, value) => context.log(`${TAG} ${pad(label)}: ${value}`);
        const warn = (message) => (context.warn || context.log).call(context, `${TAG} ${message}`);
        const error = (message) => (context.error || context.log).call(context, `${TAG} ${message}`);

        const requestId = crypto.randomUUID();
        const clientRequestId = request.headers.get('x-ms-client-request-id') || requestId;
        const headerCorrelationId = request.headers.get('x-ms-correlation-id') || null;

        context.log(`${TAG} ======== delivery received ========`);
        log('invocation', context.invocationId || requestId);

        let envelope;
        try {
            // Logged, not thrown: a missing provider setting still lets this prove decryption works.
            const absent = missingSettings(config);
            if (absent.length > 0) {
                warn(`settings not set: ${absent.join(', ')}`);
            }

            const callerAppId = readCallerAppId(request);
            log('caller appid', callerAppId || 'none (Easy Auth off, or called directly)');

            if (callerAppId && config.expectedClientId && callerAppId !== config.expectedClientId) {
                error(`caller ${callerAppId} is not ${config.expectedClientId}. ` +
                    'Easy Auth allowedApplications is not doing its job.');
                return { status: 403, jsonBody: { error: 'unexpected_caller' } };
            }

            const tokenValidation = await validateToken(request, context, requestId);
            if (!tokenValidation.ok) {
                error(`token rejected: ${tokenValidation.reason}`);
                return { status: 401, jsonBody: { error: 'unauthorized', reason: tokenValidation.reason, requestId } };
            }

            let payload;
            try {
                payload = JSON.parse(await request.text());
            } catch (parseError) {
                error(`body is not JSON: ${parseError.message}`);
                return { status: 400, jsonBody: { error: 'bad_request', reason: 'invalid JSON body', requestId } };
            }

            const parsed = parseEnvelope(payload);
            if (parsed.error) {
                error(`envelope rejected: ${parsed.error}`);
                return { status: 400, jsonBody: { error: 'bad_request', reason: parsed.error, requestId } };
            }
            envelope = parsed.envelope;

            log('type', envelope.type);
            log('tenantId', envelope.tenantId);
            log('correlationId', envelope.correlationId);
            log('channel', envelope.channel);
            log('mode', envelope.mode);
            log('ttlSeconds', envelope.ttlSeconds);

            const correlationId = envelope.correlationId || headerCorrelationId || requestId;

            // Surfaced rather than swallowed: the passcode expires before it can be used, so delivering
            // it would only produce a failed sign-in and a support call.
            if (envelope.ttlSeconds !== undefined && envelope.ttlSeconds <= 0) {
                warn(`ttlSeconds is ${envelope.ttlSeconds}; the passcode has expired.`);
            }

            let header;
            let delivery;
            try {
                ({ header, context: delivery } = await decryptDeliveryContext(
                    envelope.encryptedDeliveryContext, { config }));
            } catch (decryptError) {
                error(`decryption failed: ${decryptError.message}`);
                return { status: 400, jsonBody: { error: 'decryption_failed', correlationId, requestId } };
            }

            const kidMatches = !config.expectedKeyId || header.kid === config.expectedKeyId;
            log('kid', `${header.kid}${kidMatches ? '' : ' (DOES NOT match EPP_ENCRYPTION_KEY_ID)'}`);
            log('alg / enc', `${header.alg} / ${header.enc}`);
            log('decrypted', 'OK');
            log('nonce', delivery.nonce);

            if (config.logPlaintext) {
                // DIAGNOSTICS ONLY — writes the phone number and passcode to the log.
                log('phoneNumber', delivery.phoneNumber);
                log('extension', delivery.extension || '(none)');
                log('locale', delivery.locale);
                log('message', delivery.message);
                log('passcode', extractPasscode(delivery.message) || '(none found)');
                log('riskContext', delivery.riskContext ? JSON.stringify(delivery.riskContext) : '(none)');
            } else {
                context.log(`${TAG} plaintext suppressed (EPP_LOG_PLAINTEXT=false)`);
            }

            if (!delivery.nonce || !delivery.phoneNumber || !delivery.message) {
                error('delivery context is incomplete (nonce/phoneNumber/message)');
                return { status: 400, jsonBody: { error: 'bad_request', reason: 'incomplete delivery context', correlationId, requestId } };
            }

            const evaluation = envelope.mode === MODE.EVALUATION;
            const dispatch = contextToDispatch(delivery, envelope, clientRequestId);

            deliverInBackground(dispatch, envelope, evaluation, config, context, requestId);

            // Echoing the nonce is the whole contract: a 2xx without it is treated as a failed delivery
            // and Microsoft re-sends over its own telephony, so the user gets the code twice.
            const body = { nonce: delivery.nonce, correlationId, providerStatus: 'accepted' };

            log('responding', `200, nonce echoed, ${Date.now() - started} ms`);
            context.log(`${TAG} ======== done ========`);

            return { status: 200, jsonBody: body };
        } catch (unhandled) {
            // Verbose on purpose: this endpoint exists to diagnose onboarding.
            error(`FAILED after ${Date.now() - started} ms: ${unhandled.message}`);
            if (unhandled.cause) {
                error(`caused by: ${unhandled.cause.message || unhandled.cause}`);
            }
            context.log(`${TAG} ======== failed ========`);

            return {
                status: 500,
                jsonBody: {
                    error: 'delivery_failed',
                    detail: unhandled.message,
                    correlationId: envelope && envelope.correlationId,
                },
            };
        }
    },
});

module.exports = { whenDelivered };
